cts/cts-regression.in

```diff
@@ -1,230 +1,335 @@
+        :return: The exit code from the component's test suite.
+        :rtype: :class:`CrmExit`
+        """
+        print('Executing the %s regression tests' % self.name)
+        print('=' * 60)
+
+        cmd = self.cmd
+        if self.use_sudo and os.geteuid() != 0:
+            print('Enter the sudo password if prompted')
+            cmd = ['sudo'] + self.cmd
+
+        if verbose:
+            cmd.append('--verbose')
+
+        if self.supports_valgrind and valgrind:
+            cmd.append('--valgrind')
+
+        try:
+            rc = CrmExit(subprocess.call(cmd))
+        except OSError as err:
+            error_print('Failed to execute %s tests: %s' % (self.name, err))
+            rc = CrmExit.NOT_INSTALLED
+
+        print('=' * 60 + '\n\n')
+        return rc
+
+
+class ComponentsArgAction(argparse.Action):
+    """A class to handle `components` arguments.
+
+    This class handles special cases and cleans up the `components`
+    list. Specifically, it does the following:
+      * Enforce a default value of ['cli', 'scheduler'].
+      * Replace the 'all' alias with the components that it represents.
+      * Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
+    :meth:`parser.add_argument()` is specified, the `default` argument
+    must contain exactly one value (not `None` and not a list). We want
+    our default to be a list of components, namely `cli` and
+    `scheduler`.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        all_components = ['cli', 'exec', 'fencing', 'scheduler']
+        default_components = ['cli', 'scheduler']
+
+        if not values:
+            setattr(namespace, self.dest, default_components)
+            return
+
+        # If no argument is specified, the default gets passed as a
+        # string 'default' instead of as a list ['default']. Probably
+        # a bug in argparse. The below gives us a list.
+        if not isinstance(values, list):
+            values = [values]
+
+        components = set(values)
+
+        # If 'all', is found, replace it with the components it represents.
+        try:
+            components.remove('all')
+            components.update(set(all_components))
+        except KeyError:
+            pass
+
+        # Same for 'default'
+        try:
+            components.remove('default')
+            components.update(set(default_components))
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(components)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print.
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_components(components, verbose=False, valgrind=False):
+    """Run components' regression tests and report results for each.
+
+    :param components: A list of names of components for which to run
+        tests.
+    :type tests: list[str]
```

Comment:
Need to fix this docstring now that we're taking a list of `Component` objects and not a list of component names.